File: Python/rotatedSortedArray.py
"""Problem: 
   Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
   (i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
   You are given a target value to search. 
   If found in the array return its index, otherwise return -1.
   You may assume no duplicate exists in the array.
   Your algorithm's runtime complexity must be in the order of O(log n).
"""

def findTarget(array,target):
   low,high = 0,len(array)-1
   while low <= high:
      mid = (low+high)/2
      print("index: ",low,mid,high)
      print("value: ",array[low],array[mid],array[high])
      if array[mid] == target:
         return mid
      elif array[mid] > target:
         if array[high] > array[mid] or array[low] <= target:
            high = mid -1
         else:
            low = mid + 1
      else:
         if array[low] < array[mid] or array[high] >= target:
            low = mid + 1
         else:
            high = mid -1
   print("index: ",low,mid,high)
   return -1
   
array = [4,5,6,7,0,1,2]
print(array)
target = 3
print(target)
print(findTarget(array,target))

#Don't forget about >Binary Search

File: JS/rotatedSortedArray.js
/* This problem was asked by Amazon.

An sorted array of integers was rotated an unknown number of times.

Given such an array, find the index of the element in the array in faster than linear time. If the element doesn't exist in the array, return null.

For example, given the array [13, 18, 25, 2, 8, 10] and the element 8, return 4 (the index of 8 in the array). 

You can assume all the integers in the array are unique. */

function rotatedBinarySearch(array, target) {
  var low = 0, high = array.length-1;
  while(low <= high) {
    mid = Math.floor((low+high)/2);
    console.log('Index: ', low, mid, high);
    console.log('Value: ', array[low],array[mid],array[high]);
    if(array[mid] === target) {
      return mid;
    } else if(array[mid] > target) {
      if(array[high] > array[mid] || array[low] <= target) {
        high = mid - 1;
      } else { // (array[high] <= array[mid] || array[low] > target)
        low = mid + 1;
      }
    } else { // (array[mid] <= target)
      if(array[low] < array[mid] || array[high] >= target){
        low = mid + 1;
      } else { // (array[low] >= array[mid] || array[high] < target)
        high = mid -1;
      }
    }
  }
  return null;
}


array = [13, 18, 25, 2, 8, 10];
target = 8;
console.log('Array:', array);
console.log('Target:', target);
console.log(rotatedBinarySearch(array, target));

